### ISSUE TITLE
Allow creation of custom x509 Extensions in CSR  

### ISSUE THREAD

**Issue by nce** [pyca/cryptography#2747]
We need the possibility to add arbitrary x509 Extensions to a CSR and later allow (our) CA to sign that CSR and include these extensions in the cert.

As far as i can tell, this is currently not possible with the `x509.UnrecognizedExtension` (python 3.3.4, cryptography 1.2.2)
```
key = rsa.generate_private_key(
            public_exponent=65537,
            key_size=2048,
            backend=default_backend()
        )
custom_oid = x509.ObjectIdentifier('1.3.6.1.4.1.34380.1.1.13')
csr = x509.CertificateSigningRequestBuilder().subject_name(x509.Name([
            # Provide various details about who we are.
            x509.NameAttribute(NameOID.COUNTRY_NAME, u"US"),
            x509.NameAttribute(NameOID.STATE_OR_PROVINCE_NAME, u"CA"),
            x509.NameAttribute(NameOID.LOCALITY_NAME, u"San Francisco"),
            x509.NameAttribute(NameOID.ORGANIZATION_NAME, u"My Company"),
            x509.NameAttribute(NameOID.COMMON_NAME, u"mysite.com"),
        ])).add_extension(
                x509.UnrecognizedExtension(custom_oid, b'foobar'),
        critical=False,
        # Sign the CSR with our private key.
        ).sign(key, hashes.SHA256(), default_backend())
```
Adding the [custom extension](https://docs.puppetlabs.com/puppet/4.3/reference/ssl_attributes_extensions.html#puppet-specific-registered-ids) results in:
```
Traceback (most recent call last):
  File "/Users/foo/.pyenv/versions/3.3.4/lib/python3.3/site-packages/cryptography/hazmat/backends/openssl/backend.py", line 1552, in _create_x509_extensions
    encode = handlers[extension.oid]
KeyError: <ObjectIdentifier(oid=1.3.6.1.4.1.34380.1.1.13, name=Unknown OID)>

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
....
  File "./XXX/x509.py", line 68, in createCsr
    ).sign(key, hashes.SHA256(), default_backend())
  File "/Users/foo/.pyenv/versions/3.3.4/lib/python3.3/site-packages/cryptography/x509/base.py", line 354, in sign
    return backend.create_x509_csr(self, private_key, algorithm)
  File "/Users/foo/.pyenv/versions/3.3.4/lib/python3.3/site-packages/cryptography/hazmat/backends/multibackend.py", line 372, in create_x509_csr
    return b.create_x509_csr(builder, private_key, algorithm)
  File "/Users/foo/.pyenv/versions/3.3.4/lib/python3.3/site-packages/cryptography/hazmat/backends/openssl/backend.py", line 1349, in create_x509_csr
    gc=False
  File "/Users/foo/.pyenv/versions/3.3.4/lib/python3.3/site-packages/cryptography/hazmat/backends/openssl/backend.py", line 1555, in _create_x509_extensions
    'Extension not supported: {0}'.format(extension.oid)
NotImplementedError: Extension not supported: <ObjectIdentifier(oid=1.3.6.1.4.1.34380.1.1.13, name=Unknown OID)>
```
same exception is thrown in cryptography-1.3.dev1 (current master):
```
  File "/Users/foo/.pyenv/versions/3.3.4/lib/python3.3/site-packages/cryptography/hazmat/backends/openssl/backend.py", line 975, in _create_x509_extensions
    'Extension not supported: {0}'.format(extension.oid)
NotImplementedError: Extension not supported: <ObjectIdentifier(oid=1.3.6.1.4.1.34380.1.1.13, name=Unknown OID)>
```

We'd be glad if you could add this feature, or point us in the right direction if the error lies on our end.
So far, thanks for the great crypto plugin.

**Comment by alex** [comment]
Should be relatively straigthforward, just need to add support for `UnknownExtension` hereish: https://github.com/pyca/cryptography/blob/master/src%2Fcryptography%2Fhazmat%2Fbackends%2Fopenssl%2Fbackend.py#L971-L976

**Comment by alex** [comment]
Tentatively included in the 1.3 release
